routing/cross_mwm_index_graph.cpp

```diff
@@ -174,4 +291,84 @@ void CrossMwmIndexGraph::InsertWholeMwmTransitionSegments(NumMwmId numMwmId)
   if (!LoadWith(numMwmId, fillAllTransitionSegments))
     m_transitionCache.emplace(numMwmId, TransitionSegments());
 }
+
+void CrossMwmIndexGraph::GetBorderCross(TRoutingMappingPtr const & mapping, Segment const & s,
+                                        bool isOutgoing, vector<BorderCross> & borderCrosses)
+{
+  // ingoing edge
+  NodeIds const nodeIdsTo = GetDirectAndReverseNodeId(mapping->m_segMapping, s);
+
+  vector<ms::LatLon> const & transitionPnt =
+      isOutgoing ? GetIngoingTransitionPnt(s) : GetOutgoingTransitionPnt(s);
+  CHECK(!transitionPnt.empty(), ("Segment:", s, ", isOutgoing:", isOutgoing));
+
+  // If |isOutgoing| == true |nodes| is "to" cross nodes, otherwise |nodes| is "from" cross nodes.
+  vector<CrossNode> nodes;
+  for (ms::LatLon const & p : transitionPnt)
+    nodes.emplace_back(nodeIdsTo.m_directNodeId, nodeIdsTo.m_reverseNodeId, mapping->GetMwmId(), p);
+
+  // outgoing edge
+  vector<Segment> twins;
+  GetTwins(s, !isOutgoing, twins);
+  CHECK(!twins.empty(), ("Segment:", s, ", isOutgoing:", isOutgoing));
+  for (Segment const & twin : twins)
+  {
+    // If |isOutgoing| == true |otherSideMapping| is mapping outgoing (to) border cross.
+    // If |isOutgoing| == false |mapping| is mapping ingoing (from) border cross.
+    auto const fillBorderCrossOut = [&](TRoutingMappingPtr const & otherSideMapping) {
+      NodeIds const nodeIdsFrom = GetDirectAndReverseNodeId(otherSideMapping->m_segMapping, twin);
+      vector<ms::LatLon> const & otherSideTransitionPnt =
+          isOutgoing ? GetOutgoingTransitionPnt(twin) : GetIngoingTransitionPnt(twin);
+
+      CHECK(!otherSideTransitionPnt.empty(), ("Segment:", s, ", isOutgoing:", isOutgoing));
+      for (CrossNode const & node : nodes)
+      {
+        // If |isOutgoing| == true |otherSideNodes| is "from" cross nodes, otherwise
+        // |otherSideNodes| is "to" cross nodes.
+        BorderCross bc;
+        if (isOutgoing)
+          bc.toNode = node;
+        else
+          bc.fromNode = node;
+
+        for (ms::LatLon const & ll : otherSideTransitionPnt)
+        {
+          CrossNode & resultNode = isOutgoing ? bc.fromNode : bc.toNode;
+          resultNode = CrossNode(nodeIdsFrom.m_directNodeId, nodeIdsFrom.m_reverseNodeId,
+                                 otherSideMapping->GetMwmId(), ll);
+          borderCrosses.push_back(bc);
+        }
+      }
+    };
+    LoadWith(twin.GetMwmId(), fillBorderCrossOut);
+  }
+}
+
+CrossMwmIndexGraph::TransitionSegments const & CrossMwmIndexGraph::GetSegmentMaps(NumMwmId numMwmId)
+{
+  auto it = m_transitionCache.find(numMwmId);
+  if (it == m_transitionCache.cend())
+  {
+    InsertWholeMwmTransitionSegments(numMwmId);
+    it = m_transitionCache.find(numMwmId);
+  }
+  CHECK(it != m_transitionCache.cend(), ("Mwm ", numMwmId, "has not been downloaded."));
```

Comment:
I prefer to have a definite file and line in case of error even in a release version. |it| is used after the check and I need to know that it's valid.